sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -524,14 +635,150 @@ protected EventProcessor()
                                                        CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+            Logger.EventProcessorStop(Identifier, EventHubName, ConsumerGroup);
 
-            // TODO: Implement
+            var processingException = default(Exception);
+            var releaseSync = false;
 
-            if (async)
+            try
             {
-                await Task.Delay(1).ConfigureAwait(false);
+                // Acquire the semaphore used to synchronize processor starts and stops, respecting
+                // the async flag.  When this is held, the state of the processor is stable.
+
+                if (async)
+                {
+                    await ProcessorRunningSync.WaitAsync(cancellationToken).ConfigureAwait(false);
+                    releaseSync = true;
+                }
+                else
+                {
+                    ProcessorRunningSync.Wait();
+                    releaseSync = true;
+                }
+
+                cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+                // If the processor is not running, than it was never started or has been stopped
+                // before the semaphore was acquired; there is no work to be done.
+
+                if (_runningProcessorTask == null)
+                {
+                    return;
+                }
+
+                // Request cancellation of the running processor task.
+
+                _runningProcessorCancellationSource?.Cancel();
+                _runningProcessorCancellationSource?.Dispose();
+                _runningProcessorCancellationSource = null;
+
+                // Allow processing to complete.  If there was a processing or load balancing error,
+                // awaiting the task is where it will be surfaced.  Be sure to preserve it so
+                // that it can be surfaced.
+
+                try
+                {
+                    if (async)
+                    {
+                        await _runningProcessorTask.ConfigureAwait(false);
+                    }
+                    else
+                    {
+#pragma warning disable AZC0102 // Do not use GetAwaiter().GetResult(). Use the TaskExtensions.EnsureCompleted() extension method instead.
```

Comment:
Is there some reason we aren't using `TaskExtensions.EnsureCompleted()` here?  My understanding is that that using that by using that method we ensure we aren't actually going to end up blocking here (since in debug builds the extension method ensures that the underlying task `IsCompleted` before calling `GetAwaitier().GetResult()` and that was important in some cases to prevent weird interleaving issues.

Reply:
I'm attempting to mimic the pattern that Alex used in #10065.   My understanding from the Q&A in that PR was that the `EnsureCompleted` extension should be used in top-level, public-facing methods and that the lower-level implementation should be using `GetAwaiter().GetResult()`.    Maybe I'm misinterpreting?

I believe Pavel was the driver behind that pattern;  @pakrym, could I trouble you for your thoughts here?

Reply:
`EnsureCompleted` must be used with all the async methods that have `bool async` parameter. It guarantees that things that we expect to complete synchronously are actually synchronous.

`GetAwaiter().GetResult()` should in general be avoided but can be used if there is not other option that to run sync-over-async.

Reply:
So, just to be clear - it should be used in this context and #10065 made the initial change incorrectly?